### PR TITLE
Implementation of typed two-parted scheduler

### LINKED ISSUE #378: Restructure and reimplement `SimScheduler`
Future concept shall allow the hierarchical stacking of schedulers by separating scheduling and time advancement functionality.

Also, the activation in phases could be controlled by the time advancement unit, activating separate schedulers.

### PR DESCRIPTION
**Pull request by sebastian-peter**
Resolves #378

Some notes:
The new scheduler only stores one activation for an actor at a time.
This also means that trigger revocation happens automatically.

Adjacent changes are the following:
- CompletionMessage can only contain one trigger and should only schedule the completing actor. For other schedulings, the ScheduleTriggerMessage is used (for now)
- Removed actor reference from TriggerWithIdMessage, which has never been used

This is another step to move the scheduling mechanism in SIMONA more towards DEVS. There's more to come.

### CODE REVIEW

**Review by danielfeismann — CHANGES_REQUESTED** [review]
Thanks for your work especially for keeping ScalaDoc updated here as well. I mainly would like to double check if the todos are for the upcoming work - I least I guess so - and have some little questions.

File: src/main/scala/edu/ie3/simona/scheduler/SchedulerData.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import edu.ie3.simona.ontology.messages.SchedulerMessage
+import edu.ie3.simona.ontology.messages.SchedulerMessage.TriggerWithIdMessage
+import edu.ie3.util.scala.collection.mutable.PriorityMultiBiSet
+
+import scala.collection.mutable
+
+final case class SchedulerData(
+    parent: actor.typed.ActorRef[
+      SchedulerMessage
+    ], // FIXME typed: Scheduler or time advancer
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/scheduler/SchedulerData.scala:20]: Fixed with #652

File: src/main/scala/edu/ie3/simona/scheduler/TimeAdvancer.scala [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        nextTriggerId,
+        pauseTick
+      )
+
+    case (ctx, Stop(errorMsg: String)) =>
+      endWithFailure(ctx, data.simulation, notifier, startingTick, errorMsg)
+  }
+
+  /** TimeAdvancer is active and waiting for the current activation of the
+    * schedulee to complete
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/scheduler/TimeAdvancer.scala:100]: This was actually on purpose, meaning the actor who got scheduled (schedulee) in contrast to the scheduling actor (scheduler). However, I'm very happy to replace this term if there's a better one...

File: src/main/scala/edu/ie3/simona/api/ExtMessageUtils.scala [hunk 3]
```diff
@@ -24,16 +24,16 @@ object ExtMessageUtils {
...
   ) {
     def toSimona(triggerId: Long, triggerActor: ActorRef): CompletionMessage = {
-      val newTriggers =
-        Option.when(!extCompl.newTriggers.isEmpty) {
-          extCompl.newTriggers.asScala.map { tick =>
-            ScheduleTriggerMessage(ActivityStartTrigger(tick), triggerActor)
-          }.toSeq
+      val newTrigger =
+        // FIXME ext CompletionMessage should have only one next trigger
+        // FIXME also update the documentation in usersguide.md accordingly
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/api/ExtMessageUtils.scala:29]: Both can be solved e.g. in #652 once https://github.com/ie3-institute/simonaAPI/pull/132 is merged and released

File: src/main/scala/edu/ie3/simona/event/RuntimeEvent.scala [hunk 4]
```diff
@@ -82,7 +82,7 @@ object RuntimeEvent {
   final case class Done(
       tick: Long,
       duration: Long,
-      noOfFailedPF: Int,
+      noOfFailedPF: Int, // FIXME remove
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/event/RuntimeEvent.scala:85]: Maybe we can have a short discussion on this with next sprint meeting. In my view, the choices are 1. move the power flow statistics to `RuntimeEventListener` or 2. remove it completely

File: src/main/scala/edu/ie3/simona/ontology/messages/SchedulerMessage.scala [hunk 5]
```diff
@@ -8,30 +8,45 @@
...
     */
   case object InitSimMessage extends SchedulerMessage
 
-  /** Tell the [[SimScheduler]] to start the simulation
+  /** Starts simulation by activating the next (or first) tick
+    * @param pauseTick
+    *   Last tick that can be activated or completed before the simulation is
+    *   paused
+    *
+    * TODO rename to StartSimMessage
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/ontology/messages/SchedulerMessage.scala:33]: Yes, most of them are solved with #652 (as is this one - I wanted to keep this as short as possible, and renaming stuff can blow PRs up)

File: src/main/scala/edu/ie3/simona/ontology/messages/SchedulerMessage.scala [hunk 6]
```diff
@@ -8,30 +8,45 @@ package edu.ie3.simona.ontology.messages
...
   ) extends SchedulerMessage
 
-  /** schedule a new trigger TO the [[SimScheduler]]. This message should send
-    * only to the [[SimScheduler]]
+  /** Notifies TimeAdvancer that the simulation should stop because of some
+    * error
+    * @param errorMsg
+    *   The error message
+    *
+    * TODO only for TimeAdvancer
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/ontology/messages/SchedulerMessage.scala:44]: Solved with #652

**Review by danielfeismann — APPROVED** [review]
LGTM, thanks :)

**Pull request merged**